scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/LambdaTest.scala

```diff
@@ -0,0 +1,390 @@
+    val existingExecutions = List(StepFunctionExecution("FCL_execution_name", "a-task-token-for-fcl-task"))
+    val input = Input("TDR_execution_name", "a-task-token-for-tdr-task")
+
+    val lambdaRunResult = runLambda(input, initialDynamo, initialConfig, existingExecutions, predictableRandomNumberSelector(), Option(Errors(sendTaskSuccess = true)))
+    lambdaRunResult.result.isLeft should be(true)
+    lambdaRunResult.result.left.value.getMessage should equal("Error sending task success to step function")
+  }
+
+  "lambda" should "process tasks from existing entries in the dynamo table when no task token is passed in the input" in {
+    val initialItem = IngestQueueTableItem("TDR", Instant.now.minus(Duration.ofHours(1)), "task-token-for-tdr")
+    val initialDynamo = List(initialItem)
+    val validSourceSystems = List(SourceSystem("TDR", 2), SourceSystem("SystemTwo", 3), SourceSystem("SystemThree", 1, 100), SourceSystem("DEFAULT"))
+    val ssmParam = FlowControlConfig(6, validSourceSystems)
+    val sfnExecutions = List(StepFunctionExecution("TDR", "task-token-for-tdr"))
+    val input = Input("SOM_ExecutionName", "")
+
+    val lambdaRunResult = runLambda(input, initialDynamo, ssmParam, sfnExecutions, predictableRandomNumberSelector())
+    lambdaRunResult.result.isRight should be(true)
+    lambdaRunResult.result.getOrElse(NotImplementedError()) should be(())
+    lambdaRunResult.finalItemsInTable.size should be(0)
+    lambdaRunResult.finalStepFnExecutions.size should be(1)
+    lambdaRunResult.finalStepFnExecutions.find(_.taskToken == "task-token-for-tdr").head.taskTokenSuccess should be(true)
+  }
+
+  "lambda" should "add a new task to dynamo table and turn the task success to true when it is processed" in {
+    val initialDynamo = List.empty
+    val validSourceSystems = List(SourceSystem("TDR", 2, 25), SourceSystem("FCL", 3, 65), SourceSystem("SystemThree", 1), SourceSystem("DEFAULT", 0, 10))
+    val initialConfig = FlowControlConfig(6, validSourceSystems)
+    val existingExecutions = List(
+      StepFunctionExecution("FCL_execution_name", "a-task-token-for-fcl-task"),
+      StepFunctionExecution("TDR_execution_name", "a-task-token-for-tdr-task")
+    )
+    val input = Input("TDR_execution_name", "a-task-token-for-tdr-task")
+
+    val lambdaRunResult = runLambda(input, initialDynamo, initialConfig, existingExecutions, predictableRandomNumberSelector())
+    lambdaRunResult.result.isRight should be(true)
+    lambdaRunResult.finalStepFnExecutions.size should be(2)
+    lambdaRunResult.finalStepFnExecutions.find(_.taskToken == "a-task-token-for-fcl-task").exists(_.taskTokenSuccess) should be(false)
+    lambdaRunResult.finalStepFnExecutions.find(_.taskToken == "a-task-token-for-tdr-task").exists(_.taskTokenSuccess) should be(true)
+
+    lambdaRunResult.finalItemsInTable.size should be(0)
+  }
+
+  "lambda" should "add new task to dynamo but not send success when dedicated channel is not available for the system" in {
```

Comment:
Should this be "when dedicated channels are not available"

Reply:
we do only one task at a time. Say there are 2 dedicated channels in the config, and one of them was occupied, we will schedule a task so not sure plural will be correct?

Reply:
Ah, ok, so I guess it could be "when **a** dedicated channel is not available"